File: Library/Classes/Core/Classes/Network/CSName.m
//
// Created by Rene Dohan on 22/05/18.
// Copyright (c) 2018 Renetik Software. All rights reserved.
//

#import "CSName.h"
#import "NSString+CSExtension.h"
#import "NSMutableArray+CSExtension.h"
#import "NSObject+CSExtension.h"
#import "CSLang.h"

@implementation CSName


+ (instancetype)create:(NSString *)name :(NSString *)id {
    return [[self.class.alloc init] construct:name id:id];
}

+ (NSMutableArray<CSName *> *)createNamesFromStrings:(NSArray<NSString *> *)strings {
    NSMutableArray<CSName *> *names = NSMutableArray.new;
    var index = 0;
    for (NSString *string in strings) [names add:[CSName create:string :nil]].index = index++;
    return names;
}

- (instancetype)construct:(NSString *)name id:(NSString *)id {
    self.name = name;
    self.id = id;
    return self;
}

- (NSString *)name {
    return [self getString:@"name"];
}

- (void)setName:(NSString *)name {
    [self put:@"name" :name];
}

- (NSString *)id {
    return [self getString:@"id"];
}

- (void)setId:(NSString *)id {
    [self put:@"id" :id];
}

- (NSString *)description {
    return self.name;
}

- (BOOL)isEqual:(id)object {
    BOOL isEqual = [self.description isEqualToString:[object description]];
    if (!isEqual)return [super isEqual:object];
    return isEqual;
}


+ (CSName *)findById:(NSArray<CSName *> *)names :(NSString *)nameId {
    for (CSName *name  in names) if ([name.id equals:nameId]) return name;
    return nil;
}

+ (CSName *)findByName:(NSArray<CSName *> *)names :(NSString *)nameToFind {
    for (CSName *name  in names) if ([name.name equals:nameToFind]) return name;
    return nil;
}

@end

File: Renetik/Renetik/Classes/Core/Classes/Data/CSName.swift
//
//  CSName.swift
//  Renetik
//
//  Created by Rene Dohan on 2/20/19.
//

import RenetikObjc

open class CSName: CSDictionaryJsonData {
    public var nameKey = "name"
    public var idKey = "id"

    open var name: String {
        get { getStringValue(nameKey) }
        set(value) { put(nameKey, value) }
    }

    open var id: String {
        get { getStringValue(idKey) }
        set(value) { put(idKey, value) }
    }

    public required override init() { super.init() }

    public init(_ name: String, _ id: String = "") {
        super.init()
        self.id = id
        self.name = name
    }

    public class func create(_ name: String, _ id: String = "") -> Self {
        self.init().construct(name, id)
    }

    public class func createNames(fromStrings strings: [String]) -> [CSName] {
        var names = [CSName]()
        var index = 0
        for name: String in strings {
            names.add(create(name)).index = index
            index += 1
        }
        return names
    }

    public func construct(_ name: String, _ id: String = "") -> Self {
        self.id = id
        self.name = name
        return self
    }

    open override var description: String { name }

    open override func isEqual(_ object: Any?) -> Bool {
        if let nameObject = object as? CSName {
            return nameObject.name == name
        }
        return super.isEqual(object)
    }

    public class func find(nameId: String, _ names: [CSName]) -> CSName? {
        for name: CSName in names {
            if name.id.equals(nameId) == true {
                return name
            }
        }
        return nil
    }

    public class func find(name: String, _ names: [CSName]) -> CSName? {
        for nameObject: CSName in names {
            if nameObject.name.equals(name) {
                return nameObject
            }
        }
        return nil
    }
}
